fcrepo-exts/fcrepo-camel-toolbox
# Upgrade and Reorganize the Camel Toolbox

**Upgrade and Reorganize the Camel Toolbox**
* * *

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3755

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)
This PR won't build until https://github.com/fcrepo-exts/fcrepo-camel/pull/163 is merged.
# What does this Pull Request do?
1) Remove all karaf dependencies on fcrepo-camel-toolbox for services that are currently in highest demand:  fcrepo,fcrepo-indexing-solr and fcrepo-indexing-triplestore.
2) Upgrade camel to use the latest camel version (3.9.0)
3) Reorganize the camel toolbox so that it can be executed as a stand-alone Java 11 jar.
4) Add support for a -c <properties file path> in order to configure your camel microservices.5) Add the ability to enable select services (such as Solr or triplestore indexing) via the properties file.

# What's new?
It's an all new spring app that runs the old services without karaf. 

# How should this be tested?
Fire up Fedora 6
```
docker run -p8080:8080 -p61616:61616  -p8181:8181 --name=my_fcrepo6  fcrepo/fcrepo:6.0.0
```
Fire up SOLR  (you can use docker)
```
docker run  -p 8983:8983 --name my_solr solr:8
```
Create collection1
```
docker exec -it my_solr solr create_core -c collection1
```
Fire up fuseki
```
docker run --rm -p 3030:3030 --name my_fuseki atomgraph/fuseki --mem /test
```

Build camel toolbox
```
mvn clean install
```
Create property file using properties specified in root README of this PR.
fcepo-camel-toolbox.properties
```
solr.indexer.enabled=true
triplestore.indexer.enabled=true
triplestore.baseUrl=http://localhost:3030/test
```
Run the camel toolbox
```
java -jar fcrepo-camel-toolbox-app/target/ fcrepo-camel-toolbox-app-6.0.0-SNAPSHOT-driver.jar -c fcrepo-camel-toolbox.properties
```

Add a resource to fedora.  Verify that you can see updates to  solr and fuseki
1. open "http://localhost:8983/solr/#/collection1/query?q=*:*&q.op=OR" in a browser
2. open "http://localhost:3030/test/query?query=SELECT%20%2A%20WHERE%20%7B%20%3Fs%20%3Fo%20%3Fp%20.%20%7D"   - the query param `SELECT * WHERE { ?s ?o ?p . }` urlencoded.
3. Delete all the fedora resources and verify that all the solr and fuseki entries are gone.
4. Stop camel toolbox  and then create a bunch of resources.
5. Turn camel toolbox back on
6. Run the reindexer and verify that fuseki and solr entries have been repopulated
```
curl -v -XPOST localhost:9080/reindexing -H"Content-Type: application/json" -d '["broker:queue:solr.reindex", "broker:queue:triplestore.reindex"]'
```



# Additional Notes

Example:
* Does this change require documentation to be updated? 
* Does this change add any new dependencies? 
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? 
* Could this change impact execution of existing code?

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
